src/compiler/cse.rs

```diff
@@ -0,0 +1,452 @@
+        if info.degree < self.min_degree || info.count < self.min_occurrences {
+            0
+        } else {
+            info.count * info.degree
+        }
+    }
+}
+
+/// Common Subexpression Elimination (CSE) optimization.
+/// This optimization replaces common subexpressions with new internal signals for the step type.
+/// This is done by each time finding the optimal subexpression to replace and creating a new signal
+/// for it and replacing it in all constraints.
+/// The process is repeated until no more common subexpressions are found.
+/// Equivalent expressions are found by hashing the expressions with random assignments to the
+/// queriables. Using the Schwartz-Zippel lemma, we can determine if two expressions are equivalent
+/// with high probability.
+#[allow(dead_code)]
+pub(super) fn cse<F: Field + Hash, S: Scoring<F>>(
+    mut circuit: SBPIRMachine<F, NullTraceGenerator>,
+    config: CseConfig,
+    scorer: &S,
+) -> SBPIRMachine<F, NullTraceGenerator> {
+    for (_, step_type) in circuit.step_types.iter_mut() {
+        cse_for_step(step_type, &circuit.forward_signals, &config, scorer)
+    }
+    circuit
+}
+
+fn cse_for_step<F: Field + Hash, S: Scoring<F>>(
+    step_type: &mut StepType<F, ()>,
+    forward_signals: &[ForwardSignal],
+    config: &CseConfig,
+    scorer: &S,
+) {
+    let mut signal_factory = SignalFactory::default();
+    let mut replaced_hashes = HashSet::new();
+
+    for _ in 0..config.max_iterations {
+        let mut queriables = Vec::<Queriable<F>>::new();
+
+        forward_signals.iter().for_each(|signal| {
+            queriables.push(Queriable::Forward(*signal, false));
+            queriables.push(Queriable::Forward(*signal, true));
+        });
+        step_type.signals.iter().for_each(|signal| {
+            queriables.push(Queriable::Internal(*signal));
+        });
+
+        // Generate random assignments for the queriables
+        let mut rng = ChaCha20Rng::seed_from_u64(0);
+        let random_assignments: VarAssignments<F, Queriable<F>> = queriables
+            .iter()
+            .cloned()
+            .map(|q| (q, F::random(&mut rng)))
+            .collect();
+
+        // Turn all Expr<F, V, ()> into Expr<F, V, HashResult>
+        let mut step_type_with_hash = step_type.transform_meta(|expr| {
+            let hashed_expr = expr.hash(&random_assignments);
+            hashed_expr.meta().clone()
+        });
+
+        // Extract all the expressions from the step type
+        let mut exprs = Vec::<Expr<F, Queriable<F>, HashResult>>::new();
+
+        for constraint in &step_type_with_hash.constraints {
+            exprs.push(constraint.expr.clone());
+        }
+
+        // Find the optimal subexpression to replace
+        if let Some(common_expr) = find_optimal_subexpression(&exprs, &replaced_hashes, scorer) {
+            // Add the hash of the replaced expression to the set
+            replaced_hashes.insert(common_expr.meta().hash);
```

Comment:
Why do we need replaced_hashes? once it has been replaced, it will show only once in the constraints, and hence it will not be chosen. Right?

Reply:
Yes it is true, this was used when the replacement was being done for more than one expression at a time. Sorry